README.md
# <img src="https://uploads-ssl.webflow.com/5ea5d3315186cf5ec60c3ee4/5edf1c94ce4c859f2b188094_logo.svg" alt="Pip.Services Logo" width="200"> <br/> DataDog components for Golang

This module is a part of the [Pip.Services](http://pipservices.org) polyglot microservices toolkit.
It contains the DataDog logger and performance counters components.

The module contains the following packages:
- [**Build**](build) - contains a class used to create DataDog components by their descriptors.
- [**Clients**](clients) - contains constants and classes used to define REST clients for DataDog
- [**Count**](count) - contains a class used to create performance counters that send their metrics to a DataDog service
- [**Log**](log) - contains a class used to create loggers that dump execution logs to a DataDog service.

<a name="links"></a> Quick links:

* [Configuration](https://www.pipservices.org/recipies/configuration)
* [API Reference](https://pip-services3-node.github.io/pip-services3-datadog-node/globals.html)
* [Change Log](CHANGELOG.md)
* [Get Help](https://www.pipservices.org/community/help)
* [Contribute](https://www.pipservices.org/community/contribute)

## Use


Get the package from the Github repository:
```bash
go get -u github.com/pip-services3-go/pip-services3-datadog-go@latest
```

## Develop

For development you shall install the following prerequisites:
* Golang v1.12+
* Visual Studio Code or another IDE of your choice
* Docker
* Git

Run automated tests:
```bash
go test -v ./test/...
```

Generate API documentation:
```bash
./docgen.ps1
```

Before committing changes run dockerized test as:
```bash
./test.ps1
./clear.ps1
```

## Contacts

The library is created and maintained by **Sergey Seroukhov**.

The documentation is written by:
- **Mark Makarychev**

CHANGELOG.md
# <img src="https://uploads-ssl.webflow.com/5ea5d3315186cf5ec60c3ee4/5edf1c94ce4c859f2b188094_logo.svg" alt="Pip.Services Logo" width="200"> <br/> DataDog components for Golang Changelog

## <a name="1.1.1"></a> 1.1.1 (2023-01-12) 
- Update dependencies
## <a name="1.0.0"></a> 1.0.0 (2020-12-16) 

### Features
* Added DataDogLogClient
* Added DataDogLogger
* Added DataDogMetricsClient
* Added DataDogCounters


test/fixtures/LoggerFixture.go
package fixtures

import (
	"errors"
	"testing"
	"time"

	clog "github.com/pip-services3-go/pip-services3-components-go/log"
	"github.com/stretchr/testify/assert"
)

type LoggerFixture struct {
	logger *clog.CachedLogger
}

func NewLoggerFixture(logger *clog.CachedLogger) *LoggerFixture {
	lf := LoggerFixture{}
	lf.logger = logger
	return &lf
}

func (c *LoggerFixture) TestLogLevel(t *testing.T) {
	assert.True(t, c.logger.Level() >= clog.None)

	assert.True(t, c.logger.Level() <= clog.Trace)
}

func (c *LoggerFixture) TestSimpleLogging(t *testing.T) {
	c.logger.SetLevel(clog.Trace)

	c.logger.Fatal("", nil, "Fatal error message")
	c.logger.Error("", nil, "Error message")
	c.logger.Warn("", "Warning message")
	c.logger.Info("", "Information message")
	c.logger.Debug("", "Debug message")
	c.logger.Trace("", "Trace message")
	c.logger.Dump()

	select {
	case <-time.After(time.Duration(1000) * time.Millisecond):
	}
}

func (c *LoggerFixture) TestErrorLogging(t *testing.T) {

	var ex error = errors.New("Testing error throw")

	c.logger.Fatal("123", ex, "Fatal error")
	c.logger.Error("123", ex, "Recoverable error")
	assert.NotNil(t, ex)
	c.logger.Dump()
	select {
	case <-time.After(time.Duration(1000) * time.Millisecond):
	}
}


test/log/DataDogLogger_test.go
package log_test

import (
	"os"
	"testing"

	cconf "github.com/pip-services3-go/pip-services3-commons-go/config"
	ddlog "github.com/pip-services3-go/pip-services3-datadog-go/log"
	ddfixture "github.com/pip-services3-go/pip-services3-datadog-go/test/fixtures"

	"github.com/stretchr/testify/assert"
)

func TestDataDogLogger(t *testing.T) {
	var logger *ddlog.DataDogLogger
	var fixture *ddfixture.LoggerFixture

	apiKey := os.Getenv("DATADOG_API_KEY")
	if apiKey == "" {
		apiKey = "3eb3355caf628d4689a72084425177ac"
	}

	logger = ddlog.NewDataDogLogger()
	fixture = ddfixture.NewLoggerFixture(logger.CachedLogger)

	config := cconf.NewConfigParamsFromTuples(
		"source", "test",
		"credential.access_key", apiKey,
	)
	logger.Configure(config)

	err := logger.Open("")
	assert.Nil(t, err)

	defer logger.Close("")

	t.Run("Log Level", func(t *testing.T) {
		fixture.TestLogLevel(t)
	})

	t.Run("Simple Logging", func(t *testing.T) {
		fixture.TestSimpleLogging(t)
	})

	t.Run("Error Logging", func(t *testing.T) {
		fixture.TestErrorLogging(t)
	})

}


test/fixtures/CountersFixture.go
package fixtures

import (
	"testing"
	"time"

	ccount "github.com/pip-services3-go/pip-services3-components-go/count"
	"github.com/stretchr/testify/assert"
)

type CountersFixture struct {
	counters *ccount.CachedCounters
}

func NewCountersFixture(counters *ccount.CachedCounters) *CountersFixture {
	cf := CountersFixture{}
	cf.counters = counters
	return &cf
}

func (c *CountersFixture) TestSimpleCounters(t *testing.T) {
	c.counters.Last("Test.LastValue", 123)
	c.counters.Last("Test.LastValue", 123456)

	var counter = c.counters.Get("Test.LastValue", ccount.LastValue)
	assert.NotNil(t, counter)
	assert.NotNil(t, counter.Last)
	assert.Equal(t, counter.Last, float32(123456), 3)

	c.counters.IncrementOne("Test.Increment")
	c.counters.Increment("Test.Increment", 3)

	counter = c.counters.Get("Test.Increment", ccount.Increment)
	assert.NotNil(t, counter)
	assert.Equal(t, counter.Count, 4)

	c.counters.TimestampNow("Test.Timestamp")
	c.counters.TimestampNow("Test.Timestamp")

	counter = c.counters.Get("Test.Timestamp", ccount.Timestamp)
	assert.NotNil(t, counter)
	assert.NotNil(t, counter.Time)

	c.counters.Stats("Test.Statistics", 1)
	c.counters.Stats("Test.Statistics", 2)
	c.counters.Stats("Test.Statistics", 3)

	counter = c.counters.Get("Test.Statistics", ccount.Statistics)
	assert.NotNil(t, counter)
	assert.Equal(t, counter.Average, float32(2), 3)

	c.counters.Dump()

	select {
	case <-time.After(1000 * time.Millisecond):
		{
		}
	}
}

func (c *CountersFixture) TestMeasureElapsedTime(t *testing.T) {
	timer := c.counters.BeginTiming("Test.Elapsed")

	time.AfterFunc(100*time.Millisecond, func() {
		timer.EndTiming()

		counter := c.counters.Get("Test.Elapsed", ccount.Interval)
		assert.True(t, counter.Last > 50)
		assert.True(t, counter.Last < 5000)

		c.counters.Dump()
		select {
		case <-time.After(1000 * time.Millisecond):
			{
			}
		}
	})
}


log/DataDogLogger.go
package log

import (
	"os"
	"time"

	cconf "github.com/pip-services3-go/pip-services3-commons-go/config"
	cref "github.com/pip-services3-go/pip-services3-commons-go/refer"
	cinfo "github.com/pip-services3-go/pip-services3-components-go/info"
	"github.com/pip-services3-go/pip-services3-components-go/log"
	clog "github.com/pip-services3-go/pip-services3-components-go/log"
	clients1 "github.com/pip-services3-go/pip-services3-datadog-go/clients"
)

/**
 * Logger that dumps execution logs to DataDog service.
 *
 * DataDog is a popular monitoring SaaS service. It collects logs, metrics, events
 * from infrastructure and applications and analyze them in a single place.
 *
 * ### Configuration parameters ###
 *
 * - level:             maximum log level to capture
 * - source:            source (context) name
 * - connection:
 *     - discovery_key:         (optional) a key to retrieve the connection from [[IDiscovery]]
 *     - protocol:              (optional) connection protocol: http or https (default: https)
 *     - host:                  (optional) host name or IP address (default: http-intake.logs.datadoghq.com)
 *     - port:                  (optional) port number (default: 443)
 *     - uri:                   (optional) resource URI or connection string with all parameters in it
 * - credential:
 *     - access_key:      DataDog client api key
 * - options:
 *     - interval:        interval in milliseconds to save log messages (default: 10 seconds)
 *     - max_cache_size:  maximum number of messages stored in this cache (default: 100)
 *     - reconnect:       reconnect timeout in milliseconds (default: 60 sec)
 *     - timeout:         invocation timeout in milliseconds (default: 30 sec)
 *     - max_retries:     maximum number of retries (default: 3)
 *
 * ### References ###
 *
 * - \*:context-info:\*:\*:1.0      (optional) [[ContextInfo]] to detect the context id and specify counters source
 * - \*:discovery:\*:\*:1.0         (optional) [[IDiscovery]] services to resolve connection
 *
 * ### Example ###
 *
 *     logger := NewDataDogLogger();
 *     logger.Configure(NewConfigParamsFromTuples(
 *         "credential.access_key", "827349874395872349875493"
 *     ));
 *
 *     err := logger.Open("123");
 *
 *     logger.Error("123", ex, "Error occured: %s", ex.message);
 *     logger.Debug("123", "Everything is OK.");
 */

type DataDogLogger struct {
	*clog.CachedLogger
	client   *clients1.DataDogLogClient
	timer    chan bool
	instance string
}

//  NewDataDogLogger - Creates a new instance of the logger.
func NewDataDogLogger() *DataDogLogger {
	c := DataDogLogger{
		client: clients1.NewDataDogLogClient(nil),
	}
	c.CachedLogger = log.InheritCachedLogger(&c)
	c.instance, _ = os.Hostname()
	return &c
}

// Сonfigure -  Configures component by passing configuration parameters.
//   - config    configuration parameters to be set.
func (c *DataDogLogger) Configure(config *cconf.ConfigParams) {
	c.CachedLogger.Configure(config)
	c.client.Configure(config)
	c.instance = config.GetAsStringWithDefault("instance", c.instance)
}

// SetReferences - Sets references to dependent components.
//   - references 	references to locate the component dependencies.
func (c *DataDogLogger) SetReferences(references cref.IReferences) {
	c.CachedLogger.SetReferences(references)
	c.client.SetReferences(references)
	ref := references.GetOneOptional(cref.NewDescriptor("pip-services", "context-info", "default", "*", "1.0"))

	contextInfo, _ := ref.(*cinfo.ContextInfo)

	if contextInfo != nil && c.Source() == "" {
		c.SetSource(contextInfo.Name)
	}
	if contextInfo != nil && c.instance == "" {
		c.instance = contextInfo.ContextId
	}
}

// IsOpen - Checks if the component is opened.
//   Returns true if the component has been opened and false otherwise.
func (c *DataDogLogger) IsOpen() bool {
	return c.timer != nil
}

//  Open -  Opens the component.
//   - correlationId 	(optional) transaction id to trace execution through call chain.
//   - Returns error or nil no errors occured.
func (c *DataDogLogger) Open(correlationId string) error {
	if c.IsOpen() {
		return nil
	}

	err := c.client.Open(correlationId)
	if err == nil {
		c.timer = c.setInterval(func() { c.Dump() }, c.Interval, true)
	}

	return err

}

//  Close - Closes component and frees used resources.
//   - correlationId 	(optional) transaction id to trace execution through call chain.
//   - Returns error or nil no errors occured.
func (c *DataDogLogger) Close(correlationId string) error {
	err := c.Save(c.Cache)

	if err != nil {
		return err
	}

	if c.timer != nil {
		c.timer <- true
	}

	c.Cache = make([]*clog.LogMessage, 0, 0)
	c.timer = nil

	return c.client.Close(correlationId)

}

func (c *DataDogLogger) convertMessage(message *clog.LogMessage) clients1.DataDogLogMessage {

	tm := message.Time
	if tm.IsZero() {
		tm = time.Now().UTC()
	}
	result := clients1.DataDogLogMessage{
		Time: tm,
		Tags: map[string]string{
			"correlation_id": message.CorrelationId,
		},
		Host:    c.instance,
		Status:  clog.LogLevelConverter.ToString(message.Level),
		Message: message.Message,
	}

	result.Service = message.Source
	if result.Service == "" {
		result.Service = c.Source()
	}

	if message.Error.Message != "" || message.Error.Code != "" {
		result.ErrorKind = message.Error.Type
		result.ErrorMessage = message.Error.Message
		result.ErrorStack = message.Error.StackTrace
	}

	return result
}

//   Saves log messages from the cache.
//   - messages  a list with log messages
//   - Returns error or nil for success.
func (c *DataDogLogger) Save(messages []*clog.LogMessage) error {
	if !c.IsOpen() || len(messages) == 0 {
		return nil
	}

	data := make([]clients1.DataDogLogMessage, 0)
	for _, m := range messages {
		data = append(data, c.convertMessage(m))
	}

	return c.client.SendLogs("datadog-logger", data)
}

func (c *DataDogLogger) setInterval(someFunc func(), milliseconds int, async bool) chan bool {

	interval := time.Duration(milliseconds) * time.Millisecond
	ticker := time.NewTicker(interval)
	clear := make(chan bool)
	go func() {
		for {
			select {
			case <-ticker.C:
				if async {
					go someFunc()
				} else {
					someFunc()
				}
			case <-clear:
				ticker.Stop()
				return
			}

		}
	}()

	return clear
}


clients/DataDogLogClient.go
package clients1

import (
	"time"

	cconf "github.com/pip-services3-go/pip-services3-commons-go/config"
	cconv "github.com/pip-services3-go/pip-services3-commons-go/convert"
	cdata "github.com/pip-services3-go/pip-services3-commons-go/data"
	cerr "github.com/pip-services3-go/pip-services3-commons-go/errors"
	cref "github.com/pip-services3-go/pip-services3-commons-go/refer"
	cauth "github.com/pip-services3-go/pip-services3-components-go/auth"
	rpcclient "github.com/pip-services3-go/pip-services3-rpc-go/clients"
)

type DataDogLogClient struct {
	*rpcclient.RestClient
	defaultConfig      *cconf.ConfigParams
	credentialResolver *cauth.CredentialResolver
}

func NewDataDogLogClient(config *cconf.ConfigParams) *DataDogLogClient {

	c := &DataDogLogClient{
		RestClient:         rpcclient.NewRestClient(),
		credentialResolver: cauth.NewEmptyCredentialResolver(),
	}
	c.defaultConfig = cconf.NewConfigParamsFromTuples(
		"connection.protocol", "https",
		"connection.host", "http-intake.logs.datadoghq.com",
		"connection.port", 443,
		"credential.internal_network", "true",
	)

	if config != nil {
		c.Configure(config)
	}
	c.BaseRoute = "v1"
	return c
}

func (c *DataDogLogClient) Configure(config *cconf.ConfigParams) {
	config = c.defaultConfig.Override(config)
	c.RestClient.Configure(config)
	c.credentialResolver.Configure(config)
}

func (c *DataDogLogClient) SetReferences(refs cref.IReferences) {
	c.RestClient.SetReferences(refs)
	c.credentialResolver.SetReferences(refs)
}

func (c *DataDogLogClient) Open(correlationId string) error {
	credential, err := c.credentialResolver.Lookup(correlationId)
	if err != nil {
		return err
	}

	if credential == nil || credential.AccessKey() == "" {
		err = cerr.NewConfigError(correlationId, "NO_ACCESS_KEY", "Missing access key in credentials")
		return err
	}
	if c.Headers.Value() == nil {
		c.Headers = *cdata.NewEmptyStringValueMap()
	}
	c.Headers.SetAsObject("DD-API-KEY", credential.AccessKey())
	return c.RestClient.Open(correlationId)
}

func (c *DataDogLogClient) convertTags(tags map[string]string) string {
	if tags == nil {
		return ""
	}

	builder := ""

	for key, val := range tags {
		if builder != "" {
			builder += ","
		}
		builder += key + ":" + val
	}
	return builder
}

func (c *DataDogLogClient) convertMessage(message DataDogLogMessage) interface{} {

	timestamp := message.Time
	if timestamp.IsZero() {
		timestamp = time.Now().UTC()
	}
	result := map[string]interface{}{
		"timestamp": cconv.StringConverter.ToString(timestamp),
		"service":   message.Service,
		"message":   message.Message,
	}

	if message.Status != "" {
		result["status"] = message.Status
	} else {
		result["status"] = "INFO"
	}

	if message.Source != "" {
		result["ddsource"] = message.Source
	} else {
		result["ddsource"] = "pip-services"
	}

	if message.Tags != nil {
		result["ddtags"] = c.convertTags(message.Tags)
	}
	if message.Host != "" {
		result["host"] = message.Host
	}
	if message.LoggerName != "" {
		result["logger.name"] = message.LoggerName
	}
	if message.ThreadName != "" {
		result["logger.thread_name"] = message.ThreadName
	}
	if message.ErrorMessage != "" {
		result["error.message"] = message.ErrorMessage
	}
	if message.ErrorKind != "" {
		result["error.kind"] = message.ErrorKind
	}
	if message.ErrorStack != "" {
		result["error.stack"] = message.ErrorStack
	}

	return result
}

func (c *DataDogLogClient) convertMessages(messages []DataDogLogMessage) []interface{} {
	result := make([]interface{}, 0)

	for _, msg := range messages {
		result = append(result, c.convertMessage(msg))
	}
	return result
}

func (c *DataDogLogClient) SendLogs(correlationId string, messages []DataDogLogMessage) error {
	data := c.convertMessages(messages)

	// Commented instrumentation because otherwise it will never stop sending logs...
	//let timing = c.instrument(correlationId, "datadog.send_logs");
	result, err := c.Call(nil, "post", "input", correlationId, nil, data)
	//timing.endTiming();
	_, err = c.InstrumentError(correlationId, "datadog.send_logs", err, result)
	return err

}


clients/DataDogMetricsClient.go
package clients1

import (
	"time"

	cconf "github.com/pip-services3-go/pip-services3-commons-go/config"
	cconv "github.com/pip-services3-go/pip-services3-commons-go/convert"
	cdata "github.com/pip-services3-go/pip-services3-commons-go/data"
	cerr "github.com/pip-services3-go/pip-services3-commons-go/errors"
	cref "github.com/pip-services3-go/pip-services3-commons-go/refer"
	cauth "github.com/pip-services3-go/pip-services3-components-go/auth"
	rpcclient "github.com/pip-services3-go/pip-services3-rpc-go/clients"
)

type DataDogMetricsClient struct {
	*rpcclient.RestClient
	defaultConfig      *cconf.ConfigParams
	credentialResolver *cauth.CredentialResolver
}

func NewDataDogMetricsClient(config *cconf.ConfigParams) *DataDogMetricsClient {
	c := &DataDogMetricsClient{
		RestClient:         rpcclient.NewRestClient(),
		credentialResolver: cauth.NewEmptyCredentialResolver(),
	}

	c.defaultConfig = cconf.NewConfigParamsFromTuples(
		"connection.protocol", "https",
		"connection.host", "api.datadoghq.com",
		"connection.port", 443,
		"credential.internal_network", "true",
	)

	if config != nil {
		c.Configure(config)
	}
	c.BaseRoute = "api/v1"
	return c
}

func (c *DataDogMetricsClient) Configure(config *cconf.ConfigParams) {
	config = c.defaultConfig.Override(config)
	c.RestClient.Configure(config)
	c.credentialResolver.Configure(config)
}

func (c *DataDogMetricsClient) SetReferences(refs cref.IReferences) {
	c.RestClient.SetReferences(refs)
	c.credentialResolver.SetReferences(refs)
}

func (c *DataDogMetricsClient) Open(correlationId string) error {
	credential, err := c.credentialResolver.Lookup(correlationId)
	if err != nil {
		return err
	}

	if credential == nil || credential.AccessKey() == "" {
		err = cerr.NewConfigError(correlationId, "NO_ACCESS_KEY", "Missing access key in credentials")
		return err
	}
	if c.Headers.Value() == nil {
		c.Headers = *cdata.NewEmptyStringValueMap()
	}
	c.Headers.SetAsObject("DD-API-KEY", credential.AccessKey())
	return c.RestClient.Open(correlationId)
}

func (c *DataDogMetricsClient) convertTags(tags map[string]string) string {
	if tags == nil {
		return ""
	}

	builder := ""

	for key, val := range tags {
		if builder != "" {
			builder += ","
		}
		builder += key + ":" + val
	}
	return builder
}

func (c *DataDogMetricsClient) convertPoints(points []DataDogMetricPoint) []interface{} {

	result := make([]interface{}, 0)
	for _, p := range points {
		point := make([]string, 2)
		var tm int64 = 0
		if p.Time.IsZero() {
			tm = time.Now().UTC().UnixNano() / (int64)(time.Second)
		} else {
			tm = p.Time.UnixNano() / (int64)(time.Second)
		}
		point[0] = cconv.StringConverter.ToString(tm)
		point[1] = cconv.StringConverter.ToString(p.Value)
		result = append(result, point)
	}

	return result
}

func (c *DataDogMetricsClient) convertMetric(metric DataDogMetric) map[string]interface{} {
	tags := metric.Tags

	if metric.Service != "" {
		if tags == nil {
			tags = make(map[string]string, 0)
		}
		tags["service"] = metric.Service
	}

	result := map[string]interface{}{
		"metric": metric.Metric,
		"points": c.convertPoints(metric.Points),
	}

	if metric.Type != "" {
		result["type"] = metric.Type
	} else {
		result["type"] = "gauge"
	}

	if tags != nil {
		result["tags"] = c.convertTags(tags)
	}
	if metric.Host != "" {
		result["host"] = metric.Host
	}
	if metric.Interval > 0 {
		result["interval"] = metric.Interval
	}

	return result
}

func (c *DataDogMetricsClient) convertMetrics(metrics []DataDogMetric) map[string]interface{} {

	series := make([]interface{}, 0)
	for _, metric := range metrics {
		series = append(series, c.convertMetric(metric))
	}
	return map[string]interface{}{
		"series": series,
	}
}

func (c *DataDogMetricsClient) SendMetrics(correlationId string, metrics []DataDogMetric) error {
	data := c.convertMetrics(metrics)

	// Commented instrumentation because otherwise it will never stop sending logs...
	// timing := c.Instrument(correlationId, "datadog.send_metrics");
	result, err := c.Call(nil, "post", "series", correlationId, nil, data)
	//timing.EndTiming();
	_, err = c.InstrumentError(correlationId, "datadog.send_metrics", err, result)
	return err
}


count/DataDogCounters.go
package count

import (
	"os"

	cconf "github.com/pip-services3-go/pip-services3-commons-go/config"
	cref "github.com/pip-services3-go/pip-services3-commons-go/refer"
	ccount "github.com/pip-services3-go/pip-services3-components-go/count"
	cinfo "github.com/pip-services3-go/pip-services3-components-go/info"
	clog "github.com/pip-services3-go/pip-services3-components-go/log"
	clients1 "github.com/pip-services3-go/pip-services3-datadog-go/clients"
)

/**
 * Performance counters that send their metrics to DataDog service.
 *
 * DataDog is a popular monitoring SaaS service. It collects logs, metrics, events
 * from infrastructure and applications and analyze them in a single place.
 *
 * ### Configuration parameters ###
 *
 * - connection(s):
 *   - discovery_key:         (optional) a key to retrieve the connection from [[IDiscovery]]
 *     - protocol:            (optional) connection protocol: http or https (default: https)
 *     - host:                (optional) host name or IP address (default: api.datadoghq.com)
 *     - port:                (optional) port number (default: 443)
 *     - uri:                 (optional) resource URI or connection string with all parameters in it
 * - credential:
 *     - access_key:          DataDog client api key
 * - options:
 *   - retries:               number of retries (default: 3)
 *   - connect_timeout:       connection timeout in milliseconds (default: 10 sec)
 *   - timeout:               invocation timeout in milliseconds (default: 10 sec)
 *
 * ### References ###
 *
 * - \*:logger:\*:\*:1.0         (optional)  [[ILogger]] components to pass log messages
 * - \*:counters:\*:\*:1.0         (optional) [[ICounters]] components to pass collected measurements
 * - \*:discovery:\*:\*:1.0        (optional) [[IDiscovery]] services to resolve connection
 *
 * See [[RestService]]
 * See [[CommandableHttpService]]
 *
 * ### Example ###
 *
 *     counters := NewDataDogCounters();
 *     counters.Configure(NewConfigParamsFromTuples(
 *         "credential.access_key", "827349874395872349875493"
 *     ));
 *
 *     err := counters.Open("123",
 *         ...
 *
 *     counters.Increment("mycomponent.mymethod.calls");
 *      timing := counters.BeginTiming("mycomponent.mymethod.exec_time");
 *
 *       ...
 *
 *      timing.EndTiming();
 *
 *     counters.Dump();
 */

type DataDogCounters struct {
	*ccount.CachedCounters
	client       *clients1.DataDogMetricsClient
	logger       *clog.CompositeLogger
	opened       bool
	source       string
	instance     string
	requestRoute string
}

//  NewDataDogCounters - creates a new instance of the performance counters.
func NewDataDogCounters() *DataDogCounters {
	c := &DataDogCounters{
		client: clients1.NewDataDogMetricsClient(nil),
		logger: clog.NewCompositeLogger(),
		opened: false,
	}
	c.CachedCounters = ccount.InheritCacheCounters(c)
	c.instance, _ = os.Hostname()
	return c
}

//  Configure - configures component by passing configuration parameters.
//   - config    configuration parameters to be set.
func (c *DataDogCounters) Configure(config *cconf.ConfigParams) {
	c.CachedCounters.Configure(config)
	c.client.Configure(config)

	c.source = config.GetAsStringWithDefault("source", c.source)
	c.instance = config.GetAsStringWithDefault("instance", c.instance)
}

//  SetReferences - sets references to dependent components.
//  - references 	references to locate the component dependencies.
func (c *DataDogCounters) SetReferences(references cref.IReferences) {
	c.logger.SetReferences(references)
	c.client.SetReferences(references)
	ref := references.GetOneOptional(cref.NewDescriptor("pip-services", "context-info", "default", "*", "1.0"))

	contextInfo, _ := ref.(*cinfo.ContextInfo)

	if contextInfo != nil && c.source == "" {
		c.source = contextInfo.Name
	}
	if contextInfo != nil && c.instance == "" {
		c.instance = contextInfo.ContextId
	}
}

//  IsOpen - checks if the component is opened.
//  Returns true if the component has been opened and false otherwise.
func (c *DataDogCounters) IsOpen() bool {
	return c.opened
}

// Open - opens the component.
//   - correlationId 	(optional) transaction id to trace execution through call chain.
//   Returns  error or nil no errors occured.
func (c *DataDogCounters) Open(correlationId string) error {
	if c.opened {
		return nil
	}

	c.opened = true
	return c.client.Open(correlationId)
}

//  Close - closes component and frees used resources.
//   - correlationId 	(optional) transaction id to trace execution through call chain.
//   Returns  error or null no errors occured.
func (c *DataDogCounters) Close(correlationId string) error {
	c.opened = false

	return c.client.Close(correlationId)
}

func (c *DataDogCounters) convertCounter(counter *ccount.Counter) []clients1.DataDogMetric {
	switch counter.Type {
	case ccount.Increment:
		return []clients1.DataDogMetric{{
			Metric:  counter.Name,
			Type:    clients1.Gauge,
			Host:    c.instance,
			Service: c.source,
			Points:  []clients1.DataDogMetricPoint{{Time: counter.Time, Value: (float64)(counter.Count)}},
		}}

	case ccount.LastValue:
		return []clients1.DataDogMetric{{
			Metric:  counter.Name,
			Type:    clients1.Gauge,
			Host:    c.instance,
			Service: c.source,
			Points:  []clients1.DataDogMetricPoint{{Time: counter.Time, Value: (float64)(counter.Last)}},
		}}

	case ccount.Interval:
	case ccount.Statistics:
		return []clients1.DataDogMetric{
			{
				Metric:  counter.Name + ".min",
				Type:    clients1.Gauge,
				Host:    c.instance,
				Service: c.source,
				Points:  []clients1.DataDogMetricPoint{{Time: counter.Time, Value: (float64)(counter.Min)}},
			},
			{
				Metric:  counter.Name + ".average",
				Type:    clients1.Gauge,
				Host:    c.instance,
				Service: c.source,
				Points:  []clients1.DataDogMetricPoint{{Time: counter.Time, Value: (float64)(counter.Average)}},
			},
			{
				Metric:  counter.Name + ".max",
				Type:    clients1.Gauge,
				Host:    c.instance,
				Service: c.source,
				Points:  []clients1.DataDogMetricPoint{{Time: counter.Time, Value: (float64)(counter.Max)}},
			},
		}
	}

	return nil
}

func (c *DataDogCounters) convertCounters(counters []*ccount.Counter) []clients1.DataDogMetric {
	metrics := make([]clients1.DataDogMetric, 0)

	for _, counter := range counters {
		data := c.convertCounter(counter)

		if data != nil && len(data) > 0 {
			metrics = append(metrics, data...)
		}
	}

	return metrics
}

//   Saves the current counters measurements.
//   - counters      current counters measurements to be saves.
func (c *DataDogCounters) Save(counters []*ccount.Counter) error {
	metrics := c.convertCounters(counters)
	if len(metrics) == 0 {
		return nil
	}

	err := c.client.SendMetrics("datadog-counters", metrics)
	if err != nil {
		c.logger.Error("datadog-counters", err, "Failed to push metrics to DataDog")
	}
	return err
}
